### ISSUE TITLE
Active learning for a better SCISeg model

### ISSUE THREAD

**Issue by naga-karthik** [ivadomed/model_seg_sci#56]
We already check the box for a multi-site model, but what's even better is a multi-contrast, multi-site model! 🔥

~Currently, the model performance for sci-colorado images is better than that of sci-zurich. And, sci-colorado has axial T1w images but without the labels. We can use the existing model to do active learning on these images, manually correct the predictions and train a multi-contrast model.~

EDIT: The T1w images of `sci-colorado` are very grainy and the SC can't be seen properly. Hence, there is no point in doing active learning here. Hence, striking it out.

Another idea is to do active learning on axial T2w images of `sci-zurich`. This would be useful as it could improve the performance on the Zurich dataset.  --> **We are proceeding with this approach**

Tasks: 

- [x] Run inference on axial T2w images of `sci-zurich` 
- [x] Correct predictions to create GTs for training 
- [x] Train model with axial/sag zurich data. Check if performance on Zurich is better than just using sag T2w images. 

EDIT: updated the 3rd action task

**Comment by valosekj** [comment]
Tested on `sci-zurich` axial image (the image was not in training) on `sub-zh01_ses-01_acq-ax_T2w.nii.gz` (the subject was in training):

SC 
```
python run_inference_single_subject.py -i sci-zurich/sub-zh01/ses-01/anat/sub-zh01_ses-01_acq-ax_T2w.nii.gz -o sci-zurich/sub-zh01/ses-01/anat/sub-zh01_ses-01_acq-ax_T2w_seg.nii.gz -path-model sci-multisite-model/ -pred-type sc
```

Lesion
```
python run_inference_single_subject.py -i sci-zurich/sub-zh01/ses-01/anat/sub-zh01_ses-01_acq-ax_T2w.nii.gz -o sci-zurich/sub-zh01/ses-01/anat/sub-zh01_ses-01_acq-ax_T2w_seg_lesion.nii.gz -path-model sci-multisite-model/ -pred-type lesion
```
![sub-zh01_ses-01_acq-ax_T2w](https://github.com/ivadomed/model_seg_sci/assets/39456460/b0d4f69c-bbec-4f48-9d99-ab507b223af3)

Next step: run inference on all `sci-zurich` `acq-ax_T2w` images!

**Comment by valosekj** [comment]
### 1. Inference

I ran the inference using:

```console
sct_run_batch -config config_run_inference_sci-zurich_axial_images.json
``` 

with the following `config_run_inference_sci-zurich_axial_images.json`:

```json
{
  "path_data"   : "data/sci-zurich",
  "path_output" : "sci-zurich_inference_axial_images_2023-08-22",
  "script"      : "model_seg_sci/baselines/run_inference_sci-zurich_axial.sh",
  "jobs"        : 8,
  "script_args" : "model_seg_sci/packaging/run_inference_single_subject.py sci-multisite-model_2023-08-14"
}
```

The `run_inference_sci-zurich_axial.sh` script is added in https://github.com/ivadomed/model_seg_sci/pull/57.

### 2. QC

In the html QC report, @naga-karthik and I randomly chose five subjects (we made sure they are not in the testing split). We saved the JSON config file as `qc_artifact-sci_zurich_t2w_axial.yml` and ran the `manual_correction.py` script:

```
python manual_correction.py -path-img sci-zurich_inference_axial_images_2023-08-22/data_processed -path-label sci-zurich_inference_axial_images_2023-08-22/data_processed -config sci-zurich_inference_axial_images_2023-08-22/qc_artifact-sci_zurich_t2w_axial.yml -path-out sci-zurich -suffix-files-seg _seg_nnunet_3d -suffix-files-lesion _seg_lesion_nnunet_3d
```

The corrected SC and lesion labels were added to the `sci-zurich` dataset in the `jv/add_acq-ax_T2w_segmentations` branch.

**Comment by naga-karthik** [comment]
Just a quick recap: The current model works well on sci-colorado data but struggles with segmenting sci-zurich's lesions. There could be several reasons, i.e. mixture of both pre-op/post-op data, traumatic/ischemic lesions (both confirmed by collaborators in this [comment](https://docs.google.com/document/d/1REmxc6eiU4uqzZHqyLp-Rc1qmpZoeN7J8wcwMDCmca4/edit?disco=AAAA_LeWg5U) on ISMRM abstract), image artifacts due to metal implants, etc. The only way (according to me) to improve zurich's performance is by adding more data from Zurich. Hence, we decided to perform a [human-in-the-loop active learning](https://www.sciencedirect.com/science/article/pii/S1361841521001080) to add axial images from zurich dataset in the hopes of improving the performance. 

Toward this end, here are the latest updates:

1. We ran inference on all axial images (results/qc location: `duke/projects/ml_sci_seg/ax_zurich_inference_2023-11-03`. 
2. Randomly picked subjects by a visually assessing which predicts required the least amount of correction
3. Corrected both lesion and SC seg and uploaded to the branch `nk/add-axial-data-with-active-learning` on git-annex. Note that SC seg were _very_ good and lesion predictions considerably good with minor corrections. In this first phase, 33 subjects were corrected. 

What's next: 

1. We're thinking of strategies to add more training data (currently, subjects are randomly chosen, but we need 1-2 more ways to compare which strategy would be best)
2. A new model will be trained with the existing zurich + colorado dataset **plus** 25% of the zurich dataset's axial images (i.e. 25% of 97 subjects = 24 subjects). Why 25% --> Just a nice number to have various ablations with (i.e. 25%, 50%, 75% and so on)
